### PR TITLE
metrics: allow multiple targets for show command

### LINKED ISSUE #1497: metrics: show: support multiple targets
Currently, it only supports one single target.

From Al on discord:

```
this works fine: dvc metrics show ml_model/level_1/scores.json
this doesn't: dvc metrics show ml_model/level_{1,2}/scores.json
Error: unrecognized arguments: ml_model/level_2/scores.json
```

### PR DESCRIPTION
**Pull request by mroutis**
Fix #1497

* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [ ] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

### CODE REVIEW

File: dvc/command/metrics.py [hunk 1]
```diff
@@ -37,17 +37,21 @@ class CmdMetricsShow(CmdBase):
...
-                typ=typ,
-                xpath=xpath,
-                all_branches=self.args.all_branches,
-                all_tags=self.args.all_tags,
-                recursive=self.args.recursive,
-            )
 
-            show_metrics(metrics, self.args.all_branches, self.args.all_tags)
+        try:
+            for target in self.args.targets:
```
Comment (efiop) [inline dvc/command/metrics.py]: We should probably process all the targets we can, and only then return 1 if anything failed. E.g. see dvc/command/repro.py

Comment (mroutis) [reply dvc/command/metrics.py]: okok :+1:

**Comment by Suor** [comment]
So results are grouped by metric here, I as a user would expect them to be grouped by experiment.

**Comment by efiop** [comment]
@Suor What is "experiment"?

**Comment by Suor** [comment]
@efiop I mean branch, tag, etc

File: dvc/version.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 import subprocess
 
 
-_BASE_VERSION = "0.53.2"
```
Comment (efiop) [inline dvc/version.py]: Looks like something went wrong during your rebase :slightly_smiling_face:

File: dvc/repo/metrics/show.py [hunk 3]
```diff
@@ -270,14 +279,16 @@ def show(
...
+        )
         metrics = _read_metrics(repo, entries, branch)
         if metrics:
             res[branch] = metrics
 
     if not res:
-        if path:
-            raise BadMetricError(path)
+        if targets:
+            raise BadMetricError(targets[0])
```
Comment (efiop) [inline dvc/repo/metrics/show.py]: Maybe we should somehow adjust this error? It seems weird to use targets[0].

File: tests/func/test_metrics.py [hunk 4]
```diff
@@ -711,7 +713,7 @@ def _test_metrics(self, func):
         )
 
         res = self.dvc.metrics.show(
-            "", all_branches=True, typ="json", xpath="metrics"
+            [], all_branches=True, typ="json", xpath="metrics"
```
Comment (efiop) [inline tests/func/test_metrics.py]: I guess we could simply
```suggestion
            all_branches=True, typ="json", xpath="metrics"
```
here :slightly_smiling_face:

File: dvc/repo/metrics/show.py [hunk 5]
```diff
@@ -171,20 +172,28 @@ def _collect_metrics(repo, path, recursive, typ, xpath, branch):
...
     outs = [out for stage in repo.stages() for out in stage.outs]
+    found = []
 
-    if path:
+    # Display all metrics if no target is given
+    if not targets:
+        targets = []
+        found = outs
+
+    for target in targets:
```
Comment (efiop) [inline dvc/repo/metrics/show.py]: This feels weird, this `for` is disconnected from `if not targets` above. How about something like
```
if targets:
    # maybe even move
    for target in targets:
    ....
else:
    found = outs
```

File: dvc/repo/metrics/show.py [hunk 6]
```diff
@@ -171,20 +172,28 @@ def _collect_metrics(repo, path, recursive, typ, xpath, branch):
...
+                target, outs=outs, recursive=recursive
+            )
         except OutputNotFoundError:
             logger.debug(
                 "DVC-file not for found for '{}' in branch '{}'".format(
-                    path, branch
+                    target, branch
                 )
             )
             return []
```
Comment (efiop) [inline dvc/repo/metrics/show.py]: Also this no longer makes sense, because you are returning on first failed target, even though you should process everything you can.

File: dvc/command/metrics.py [hunk 7]
```diff
@@ -121,7 +119,10 @@ def add_parser(subparsers, parent_parser):
         formatter_class=argparse.RawDescriptionHelpFormatter,
     )
     metrics_show_parser.add_argument(
-        "path", nargs="?", help="Path to a metric file or a directory."
+        "targets",
+        nargs="*",
+        help="Metric files or directories to show "
```
Comment (shcheklein) [inline dvc/command/metrics.py]: `directories (see `-R`)` ?

File: dvc/repo/metrics/show.py (hunk 6 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: Maybe it's better to raise the `BadMetricError` instead.

Comment (mroutis) [reply dvc/repo/metrics/show.py]: Never mind, it halts the metric recollection with `all_branches` / `all_targets`.

File: dvc/repo/metrics/show.py [hunk 8]
```diff
@@ -152,15 +152,16 @@ def _read_metric(fd, typ=None, xpath=None, fname=None, branch=None):
...
+        targets (list): List of paths to metric files or directories.
+            If there's no target, collect all metrics.
         recursive (bool): If path is a directory, do a recursive search for
             metrics on the given path.
-        typ (str): The type of metric to search for, could be one of the
-            following (raw|json|tsv|htsv|csv|hcsv).
+        typ (str): The type that will be used to interpret the metric file,
+            one of the followings - (raw|json|tsv|htsv|csv|hcsv).
         xpath (str): Path to search for.
         branch (str): Branch to look up for metrics.
```
Comment (efiop) [inline dvc/repo/metrics/show.py]: Btw, deepsource reports that `branch` is not used here. Let's remove it while we are at it.

File: dvc/repo/metrics/show.py [hunk 9]
```diff
@@ -171,20 +172,21 @@ def _collect_metrics(repo, path, recursive, typ, xpath, branch):
             - xpath:
     """
     outs = [out for stage in repo.stages() for out in stage.outs]
+    found = []
 
-    if path:
-        try:
-            outs = repo.find_outs_by_path(path, outs=outs, recursive=recursive)
-        except OutputNotFoundError:
-            logger.debug(
```
Comment (efiop) [inline dvc/repo/metrics/show.py]: why did you remove debug message though?

Comment (mroutis) [reply dvc/repo/metrics/show.py]: :thinking: thought you said we don't need this anymore.
https://github.com/iterative/dvc/pull/2319#discussion_r310354892

Comment (mroutis) [reply dvc/repo/metrics/show.py]: my bad, then, I'll bring it back

File: dvc/repo/metrics/show.py [hunk 10]
```diff
@@ -269,14 +271,16 @@ def show(
...
+        )
         metrics = _read_metrics(repo, entries, branch)
         if metrics:
             res[branch] = metrics
 
     if not res:
-        if path:
-            raise BadMetricError(path)
+        if targets:
+            raise BadMetricError(targets)
```
Comment (efiop) [inline dvc/repo/metrics/show.py]: So this error will be printed only if all targets failed, right? This feels wrong.

File: dvc/repo/metrics/show.py (hunk 9 above)
Comment (efiop) [reply dvc/repo/metrics/show.py]: @mroutis I meant that `return []` doesn't make sense, because we should try to process all targets.

File: dvc/repo/metrics/show.py [hunk 11]
```diff
@@ -171,20 +172,26 @@ def _collect_metrics(repo, path, recursive, typ, xpath, branch):
...
                 )
-            )
-            return []
+            except OutputNotFoundError:
+                logger.debug(
+                    "DVC-file not for found for '{}' in branch '{}'".format(
+                        target, branch
+                    )
+                )
+                pass
```
Comment (efiop) [inline dvc/repo/metrics/show.py]:
```suggestion
```

File: dvc/repo/metrics/show.py (hunk 10 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: Yes, only if the targets didn't show any result.

Scenarios:
  * You have a branch `master` (which doesn't have any content) then a branch `experiment-1` (which contains a `metrics.json` file). If you do `dvc metrics show --all-branches it should show only that branch without master.
```
experiment-1:
    metrics.json: { AUC: 3}
```
* If you have two branches, one with `metrics_1.json` and the other one with `metrics_1.json` and `metrics_2.json`, if you do `dvc metrics show --all-branches metrics_{1,2,3}.json` it should show the 1 and 2.
Maybe we should print an error like `metric_3.json` was not found. Would it feel less wrong in such case?

Comment (efiop) [reply dvc/repo/metrics/show.py]: Sorry for such ^ an open-ended comment :) I am wondering, if, for example, we pass `metric1 metric2` to `metrics show`, and there is nothing for `metric2` but there is something for `metric1`, should this be an error? Feels like yes. And from CLI standpoint, it should process everything it can, show it and then finish with non-zero error code if anything went wrong. The obvious way to solve it is to continue processing targets one-by-one as we used to and group them(merge returned dicts) in CLI part, catching BadMetricError's and returning proper error code. What do you think?

Comment (mroutis) [reply dvc/repo/metrics/show.py]: I like the idea of using error codes to show if a metric was found or not.

So you think about moving the `for target in targets` to the command itself and leave `repo.metrics.show` to accept a single path instead of a list of targets?

Comment (efiop) [reply dvc/repo/metrics/show.py]: @mroutis That would be the most straightforward way to do that, yes. There might be another way to achieve both, though. For example, in the returned dict where metrics are sorted by branch, we could introduce a special key (maybe `None`?) that would contain a list of targets that were not found anywhere, so we could use that in CLI to return an error. Btw, that might be a good way to report missing ones too, e.g.

```
$ dvc metrics show auc.metric something

baseline-experiment:
    auc.metric: 0.588426
bigram-experiment:
    auc.metric: 0.602818

The following outputs do not exist, are not metric files or are malformed: 
    something
```
What do you think? :)

Comment (mroutis) [reply dvc/repo/metrics/show.py]: great :+1:

File: dvc/command/metrics.py [hunk 12]
```diff
@@ -35,19 +36,35 @@ def show_metrics(metrics, all_branches=False, all_tags=False):
...
+            missing = set(self.args.targets) - found
+
+            if missing:
+                logger.error(
+                    "the following metrics do not exists, "
+                    "are not metric files or are malformed: {paths}".format(
+                        paths=", ".join("'{}'".format(x) for x in missing)
+                    )
+                )
+                return 1
```
Comment (pared) [inline dvc/command/metrics.py]: I would suggest to write test for this case.

File: dvc/command/metrics.py [hunk 13]
```diff
@@ -46,6 +47,24 @@ def run(self):
...
 
             show_metrics(metrics, self.args.all_branches, self.args.all_tags)
+
+            found = set(
+                itertools.chain.from_iterable(
+                    metric.keys() for metric in metrics.values()
+                )
+            )
+
+            missing = set(self.args.targets) - found
```
Comment (efiop) [inline dvc/command/metrics.py]: and what if target was a directory with -R? You won't find it in keys.

Comment (mroutis) [reply dvc/command/metrics.py]: :thinking: that's true!
